community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/spi/ProcedureSignature.scala

```diff
@@ -19,33 +19,43 @@
  */
 package org.neo4j.cypher.internal.compiler.v3_0.spi
 
+import java.util
+
 import org.neo4j.cypher.internal.compiler.v3_0.executionplan.{InternalQueryType, READ_ONLY, READ_WRITE}
 import org.neo4j.cypher.internal.frontend.v3_0.symbols.CypherType
 
-sealed trait ProcedureMode {
+import scala.collection.JavaConverters._
+
+sealed trait ProcedureCallMode {
   val queryType: InternalQueryType
 
   def call(ctx: QueryContext, signature: ProcedureSignature, args: Seq[Any]): Iterator[Array[AnyRef]]
 }
 
-case object ProcReadOnly extends ProcedureMode {
+case object LazyReadOnlyCallMode extends ProcedureCallMode {
   override val queryType: InternalQueryType = READ_ONLY
 
   override def call(ctx: QueryContext, signature: ProcedureSignature, args: Seq[Any]): Iterator[Array[AnyRef]] =
-    ctx.callReadOnlyProcedure(signature, args)
+    ctx.callReadOnlyProcedure(signature.name, args)
 }
 
-case object ProcReadWrite extends ProcedureMode {
+case object EagerReadWriteCallMode extends ProcedureCallMode {
   override val queryType: InternalQueryType = READ_WRITE
 
-  override def call(ctx: QueryContext, signature: ProcedureSignature, args: Seq[Any]): Iterator[Array[AnyRef]] =
-    ctx.callReadWriteProcedure(signature, args)
+  override def call(ctx: QueryContext, signature: ProcedureSignature, args: Seq[Any]): Iterator[Array[AnyRef]] = {
+    val buffer = new util.ArrayList[Array[AnyRef]]()
```

Comment:
Since we turn this back into a scala iterator right away, could we not use an `ArrayBuffer` instead.